Update conda-store docs (#399)
nebari-dev/nebari-docs: docs/docs/tutorials/creating-new-environments.md

---
id: creating-new-environments
title: Manage packages and environments
description: Using conda-store for environment management
---

Nebari uses [`conda-store`][conda-store-docs] for managing and sharing reproducible environments (collection of specific packages and versions) on the platform.

:::warning
conda-store is the most reliable way to manage your packages and environments on Nebari.

> Do not install libraries directly in the notebook or through the terminal (outside a conda-store managed environment). It won't work as expected and may break your instance in subtle ways.
> :::

Reviewer: ```suggestion
Please refrain from installing libraries directly in the notebook or through the terminal (outside a conda-store managed environment). These actions may cause unexpected issues with your environment in subtle ways. 
:::
```
Maybe we can reprhsae this a bit, what do you think?

## Preliminary reading

It's useful to understand basics of conda-store and how it builds on top of the conda ecosystem, to use it effectively in Nebari.

- [Essential conda concepts: Packages, environments, channels, dependencies, etc.](https://conda.store/conda-store/explanations/conda-concepts)
- [conda-store concepts: Reproducibility, namespaces, environment versions, roles, etc.](https://conda.store/conda-store/explanations/conda-store-concepts)

Nebari has conda-store integrated, and you can use it through the graphical UI.

## Open `conda-store` web interface

> To navigate to conda-store:

Reviewer 2: ```suggestion
There are several options for navigating to conda-store:
```
It wasn't immediately clear that this was a list of options

> * From Nebari Home, click on **"Environment Management"** under "Service"

Reviewer 2: ```suggestion
* From Nebari Home, click on **"Environment Management"** under "Services"
```
"Nebari Home" is kind of ambiguous. I'm assuming you mean the App Launcher? An image may help here. 

Also, the App Launcher is optional so should we point that out in this option

* From JupyterLab, click on `Nebari` in the menu bar and go to **"Environments"**
* From anywhere, go to URL: `https://<your-nebari-domain>/conda-store`

If not logged in to conda-store, click on the **"Log in"** button in th left sidebar and authenticate similar to the [Nebari login][login-keycloak]. This is required to be able to access many `conda-store` features.

## Create, edit, and manage environments

Go through the following conda-store (UI) tutorials
on using the graphical interface for various actions:

* [Create new environments][cs-create-env]
* [Edit & delete existing environments][cs-edit-delete-env]
* [Switch environment versions][version-control]

## Default namespaces in Nebari

A default Nebari deployment/instance has the following namespaces corresponding to [Nebari groups][configure-keycloak-groups]:

<!-- Verify the roles and actions -->

* `analyst` namespace - Users in the `analyst` group can view and `admin` group can view+edit the environments in this namespace

> * `developer` namespace - Users in the `developer` and `admin` groups can view+edit the environments in this namespace

Reviewer 2: I'm pretty sure developer is required for access to the dask workers.

Reviewer: I was about to suggest some changes, but noticed this will be confusing with the keycloak roles and the concepts of namespaces intertwined, so I will think about how to rewrite this a bit

* `nebari-git` namespace - Everyone can view and `admin`s can edit

As an individual user, you also have a personal namespace with the same name as your Nebari username.

:::note
If you can "view" an environment, you can use it.
:::

## Select environments in editors

Instructions to select any environment you have access to in the following editing spaces:

* **JupyterLab** - In a Jupyter Notebook, click on the "Select Kernel" dropdown in the top-left corner, and select the environment.

* **VS Code** - Click on the ⚙️ icon in the bottom-right to open `Settings` -> `Command Pallette`, and type "Python: Select Interpreter" and press <kbd>Enter</kbd> to get the list of environments to select from.

* **Terminal** - In the terminal window, you can use `conda` CLI commands like `conda activate <namespace>-<environment_name>` to activate the relevant environment and `conda env list` to view the list of available environments.

## Special requirements

### Dask

> Include the [`nebari-dask` metapackage](https://anaconda.org/conda-forge/nebari-dask) in your environment to use Dask. This ensures you have the correct version of `dask-gateway` and the latest versions of `dask` and `distributed` libraries.

Reviewer 2: Add note that the nebari-dask metapackage version should match nebari version

Reviewer 2: also note again here that you have to be in the develop group to be able to access dask.

By default, the `nebari-git-nebari-git-dask` environment (available to everyone) can be used for basic Dask workflows.

### JHub App Launcher

> Include the `jhub-apps` package in your environment to create apps using the JHub App Launcher. You will also need the relevant app development framework and other necessary packages in the environment.

Reviewer 2: You don't need jhub-apps itself. You need `jhsingle-native-proxy>=0.8.2` for all apps. 

For panel apps you'll also need `bokeh-root-cmd>=0.1.2` in additional to the panel deps.

Reviewer 2: The app launcher now has its own docs which explain the environment requirements. I think its safe to remove this here.
